main/src/library/LazyList.flix

```diff
@@ -0,0 +1,275 @@
+            else if (n == 1) LazyCons(h, lazy Empty)
+            else LazyCons(h, lazy take(n - 1, force t))
+    }
+
+    ///
+    /// Returns `xs` without the first `n` elements.
+    ///
+    /// Forces the first `n` elements of `xs`.
+    ///
+    pub def drop(n: Int32, xs: LazyList[a]): LazyList[a] = match xs {
+        case Empty => xs
+        case LazyCons(_, t) => if (n == 0) xs else drop(n - 1, force t)
+    }
+
+    ///
+    /// Returns `x` without the first element.
+    ///
+    /// Forces the first element of `xs`.
+    ///
+    pub def tail(xs: LazyList[a]): LazyList[a] = drop(1, xs)
+
+    ///
+    /// Returns `xs` as a `List`.
+    ///
+    /// Forces the entire list `xs`.
+    ///
+    pub def toList(xs: LazyList[a]): List[a] =
+        toListHelper(xs, Nil)
+
+    pub def toListHelper(xs: LazyList[a], acc: List[a]): List[a] = match xs {
+        case Empty => List.reverse(acc)
+        case LazyCons(h, t) => toListHelper(force t, h :: acc)
+    }
+
+    ///
+    /// Alias for `foldLeft`.
+    ///
+    /// Forces the entire list `xs`.
+    ///
+    pub def fold(f: (b, a) -> b, s: b, xs: LazyList[a]): b = foldLeft(f, s, xs)
+
+    ///
+    /// Applies `f` to a start value `s` and all elements in `xs` going from left to right.
+    ///
+    /// That is, the result is of the form: `f(...f(f(s, x1), x2)..., xn)`.
+    ///
+    /// Forces the entire list `xs`.
+    ///
+    pub def foldLeft(f: (b, a) -> b, s: b, xs: LazyList[a]): b = match xs {
+        case Empty => s
+        case LazyCons(h, t) => foldLeft(f, f(s, h), force t)
+    }
+
+    ///
+    /// Applies `f` to a start value `s` and all elements in `xs` going from right to left.
+    ///
+    /// That is, the result is of the form: `f(x1, ...f(xn-1, f(xn, s))...)`.
+    ///
+    /// Forces the entire list `xs`.
+    ///
+    pub def foldRight(f: (a, Lazy[b]) -> b, s: b, xs: LazyList[a]): b = match xs {
+        case Empty => s
+        case LazyCons(h, t) => f(h, lazy foldRight(f, s, force t))
+    }
+
+    ///
+    /// Returns `true` if and only if at least one element in `xs` satisfies the predicate `f`.
+    ///
+    /// Returns `false` if `xs` is empty.
+    ///
+    /// Forces the list `xs` until an element that satisfies the predicate `f` is found.
+    ///
+    pub def exists(f: a -> Bool, xs: LazyList[a]): Bool = match xs {
+        case Empty => false
+        case LazyCons(h, t) => f(h) or exists(f, force t)
+    }
+
+    ///
+    /// Returns `true` if and only if all elements in `xs` satisfy the predicate `f`.
+    ///
+    /// Returns `true` if `xs` is empty.
+    ///
+    /// Forces the list `xs` until an element that does not satisfy the predicate `f` is found.
+    ///
+    pub def forall(f: a -> Bool, xs: LazyList[a]): Bool = match xs {
+        case Empty => true
+        case LazyCons(h, t) => f(h) and forall(f, force t)
+    }
+
+    ///
+    /// Returns the result of applying `f` to every element in `xs`.
+    ///
+    /// Forces the entire list `xs`.
```

Comment:
But mapping does not force the list!

Reply:
You are right. I think I misunderstood here.